docs/RFCS/20191009_gc_protected_timestamps.md

```diff
@@ -0,0 +1,597 @@
+pkg/protectedts/tracker - implements protectedtsbase.Tracker on top of Catalog
+```
+
+### Implementation of Store and Catalog
+
+This RFC chooses to store the state in SQL system tables (in contrast to
+KVs with their own ad-hoc structure). Futhermore the approach taken was to
+utilize SQL for data modeling rather than to store encoded protocol buffers.
+
+The protected timestamp subsystem add 3 system tables (though one is interleaved
+in another).
+
+The current state of the table is stored in a special table called
+`system.protected_ts_meta`. 
+```sql
+  CREATE TABLE protected_ts_meta (
+      version INT8 NOT NULL,
+      rows    INT4 NOT NULL,
+      spans   INT4 NOT NULL,
+      FAMILY "primary" (_exists, version, rows, spans),
+      CONSTRAINT check_exists CHECK (_exists)
+  )
+```
+
+What doesn't show up in this is a hidden columns to ensure that this table has
+one row. The hidden column is defined as:
+
+```
+_exists BOOL NOT NULL CHECK (_exists) DEFAULT (true) PRIMARY KEY
+```
+
+The other tables are defined as:
+
+```sql 
+  CREATE TABLE protected_ts_timestamps (
+      id UUID NOT NULL,
+      ts DECIMAL NOT NULL,
+      mode INT NOT NULL,
+      meta_type STRING NOT NULL,
+      meta BYTES NULL,
+      CONSTRAINT "primary" PRIMARY KEY (id ASC),
+      FAMILY "primary" (id, ts, meta_type, meta)
+  )
+
+  CREATE TABLE protected_ts_spans (
+    id UUID NOT NULL,
+    key BYTES NOT NULL,
+    end_key BYTES NOT NULL,
+    CONSTRAINT "primary" PRIMARY KEY (id ASC, key ASC, end_key ASC),
+    CONSTRAINT fk_protected_ts_timestamps FOREIGN KEY (id) REFERENCES protected_ts_timestamps(id) ON DELETE CASCADE,
+    FAMILY "primary" (id, key, end_key)
+  ) INTERLEAVE IN PARENT protected_ts_timestamps (id)
+```
+
+The implementation provided by `protectedts/provider` interact with this
+schema using a `sqlutils.InternalExecutor`.
+
+### Implementation of `Tracker`
+
+The tracker which is utilized by the `storage` package wraps a provider and
+will periodically poll the state from the catalog based on a cluster setting.
+
+The polling mechanism will initialize by reading the entire state and observing
+the commit timestamp. Later attempts will begin by just retrieving the metadata
+and only retrieving the full state when the version changes.
+
+Internally the tracker could use a data structure like an interval tree to make
+answering queries cheap. Given the small size of the data set however there are
+likely simpler solutions.
+
+## Drawbacks
+
+May extend the life of bytes marginally, especially when the GC TTL is very
+short (seconds).
+
+## Future work
+
+### More exact protection of the desired time
+
+In theory a protected timestamp does not need to protect all versions alive at
+and newer than some timestamp but rather could just protect the version alive
+at exactly that timestamp.
+
+For example imagine the following history for key `k`:
+
+```
+  5: put baz
+  4: put bar
+  3: --- protected ---
+  2: delete
```

Comment:
Is it important in this example that `delete` is right after the protected ts? Doesn't this allow us to remove `k@2` during the GC? Maybe not because we want a snapshot of the latest change to `k`? Is this handling of deletion tombstones after the GC threshold discussed at all in this RFC?